File: app/src/main/java/za/co/woolworths/financial/services/android/ui/fragments/wtoday/WTodayFragment.java
package za.co.woolworths.financial.services.android.ui.fragments.wtoday;

import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Message;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.awfs.coordination.BR;
import com.awfs.coordination.R;
import com.awfs.coordination.databinding.WtodayFragmentBinding;

import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties;
import za.co.woolworths.financial.services.android.models.WoolworthsApplication;
import za.co.woolworths.financial.services.android.ui.base.BaseFragment;
import za.co.woolworths.financial.services.android.util.ErrorHandlerView;
import za.co.woolworths.financial.services.android.util.Utils;

public class WTodayFragment extends BaseFragment<WtodayFragmentBinding, WTodayViewModel> implements WTodayNavigator {

	private WTodayViewModel wTodayViewModel;
	private WebView webView;
	private ErrorHandlerView mErrorHandlerView;
	private String TAG = this.getClass().getSimpleName();

	@Override
	public WTodayViewModel getViewModel() {
		return wTodayViewModel;
	}

	@Override
	public int getBindingVariable() {
		return BR.viewModel;
	}

	@Override
	public int getLayoutId() {
		return R.layout.wtoday_fragment;
	}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		wTodayViewModel = ViewModelProviders.of(this).get(WTodayViewModel.class);
		wTodayViewModel.setNavigator(this);
		try {
			hideToolbar();
		} catch (NullPointerException ex) {
			Log.e(TAG, ex.toString());
		}
	}

	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		Utils.updateStatusBarBackground(getActivity());
		Activity activity = getActivity();
		if (activity != null) {
			hideToolbar();
			mErrorHandlerView = new ErrorHandlerView(activity, getViewDataBinding().incNoConnectionHandler.noConnectionLayout);
			mErrorHandlerView.setMargin(getViewDataBinding().incNoConnectionHandler.noConnectionLayout, 0, 0, 0, 0);
			initWebView(view);
			retryConnect(view);
		}
	}

	@Override
	public void onResume() {
		super.onResume();
		Utils.setScreenName(getActivity(), FirebaseManagerAnalyticsProperties.ScreenNames.WTODAY);
	}

	@SuppressLint("SetJavaScriptEnabled")
	private void initWebView(View view) {
		webView = view.findViewById(R.id.wtoday_fragment_webview);
		webView.getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
		webView.getSettings().setJavaScriptEnabled(true);
		webView.getSettings().setMediaPlaybackRequiresUserGesture(false);
		webView.getSettings().setDomStorageEnabled(true);
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
			webView.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE);
		webView.addJavascriptInterface(new WebAppInterface(getActivity()), "Android");
		webView.loadUrl(WoolworthsApplication.getWwTodayURI());
		webView.getSettings().setSupportMultipleWindows(true);
		webView.setWebViewClient(new WebViewClient() {
			@TargetApi(android.os.Build.VERSION_CODES.M)
			@Override
			public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
				super.onReceivedError(view, request, error);
				mErrorHandlerView.webViewBlankPage(view);
				mErrorHandlerView.networkFailureHandler(error.toString());
			}

			@Override
			public boolean shouldOverrideUrlLoading(WebView view, String url) {
				return super.shouldOverrideUrlLoading(view, url);
			}

			@SuppressWarnings("deprecation")
			@Override
			public void onReceivedError(WebView webView, int errorCode, String description, String failingUrl) {
				mErrorHandlerView.webViewBlankPage(webView);
				mErrorHandlerView.networkFailureHandler(description);
			}
		});

		webView.setWebChromeClient(new WebChromeClient(){
			@Override
			public boolean onCreateWindow(WebView view, boolean isDialog, boolean isUserGesture, Message resultMsg) {
				String data = view.getHitTestResult().getExtra();
				try {
					if (!TextUtils.isEmpty(data)) {
						Uri uri = Uri.parse(data);
						Intent intent = new Intent(Intent.ACTION_VIEW, uri);
						if (intent.resolveActivity(getActivity().getPackageManager()) != null) {
							startActivity(intent);
						}
					}
				} catch (Exception e) {
					Log.e(TAG, e.getMessage());
				}
				return false;
			}
		});
		
	}

	private void retryConnect(View view) {
		view.findViewById(R.id.btnRetry).setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				if (isNetworkConnected()) {
					mErrorHandlerView.hideErrorHandler();
					webView.loadUrl(WoolworthsApplication.getWwTodayURI());
				}
			}
		});
	}

	@Override
	public void onHiddenChanged(boolean hidden) {
		super.onHiddenChanged(hidden);
		if (!hidden) {
			//do when hidden
			setStatusBarColor(R.color.white);
			hideToolbar();
		}
	}

	public void scrollToTop() {
		ObjectAnimator anim = ObjectAnimator.ofInt(webView, "scrollY", webView.getScrollY(), 0);
		anim.setDuration(500).start();
	}
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/fragments/wtoday/WTodayFragment.kt
package za.co.woolworths.financial.services.android.ui.fragments.wtoday

import android.Manifest
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Message
import android.util.Log
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.webkit.*
import android.webkit.WebViewClient.ERROR_CONNECT
import android.webkit.WebViewClient.ERROR_TIMEOUT
import androidx.core.content.ContextCompat
import com.awfs.coordination.databinding.WtodayMainFragmentBinding
import dagger.hilt.android.AndroidEntryPoint
import za.co.woolworths.financial.services.android.common.notificationpermission.NotificationPermission
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties
import za.co.woolworths.financial.services.android.contracts.IWTodayInterface
import za.co.woolworths.financial.services.android.models.dto.ProductsRequestParams
import za.co.woolworths.financial.services.android.ui.activities.dashboard.BottomNavigationActivity
import za.co.woolworths.financial.services.android.ui.extension.isConnectedToNetwork
import za.co.woolworths.financial.services.android.ui.extension.isEmailValid
import za.co.woolworths.financial.services.android.ui.fragments.product.grid.ProductListingFragment
import za.co.woolworths.financial.services.android.util.PermissionResultCallback
import za.co.woolworths.financial.services.android.util.PermissionUtils
import za.co.woolworths.financial.services.android.util.Utils
import javax.inject.Inject

@Suppress("DEPRECATION")
@AndroidEntryPoint
class WTodayFragment : WTodayExtension(), IWTodayInterface, PermissionResultCallback {

    @Inject
    lateinit var notificationPermission: NotificationPermission
    private lateinit var permissionUtils: PermissionUtils

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activity?.let {
            Utils.updateStatusBarBackground(it)
            permissionUtils = PermissionUtils(it, this@WTodayFragment)
        }


    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.configureUI()
        binding.setClient()
        checkRunTimePermissionForLocation()
        askNotificationPermission()
    }

    fun checkRunTimePermissionForLocation(): Boolean {
        permissionUtils?.apply {
            val permissions = ArrayList<String>()
            permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
            return checkAndRequestPermissions(permissions, 3)
        }
        return false
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun WtodayMainFragmentBinding.configureUI() {
        webWToday?.apply {
            settings?.apply {
                javaScriptEnabled = true
                mediaPlaybackRequiresUserGesture = true
                domStorageEnabled = true
                cacheMode = WebSettings.LOAD_NO_CACHE
                addJavascriptInterface(WebViewJavascriptInterface(this@WTodayFragment), "Android")
                setSupportMultipleWindows(true)

                mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
            }

            loadUrl(wTodayUrl)
        }

        incNoConnectionHandler.btnRetry?.setOnClickListener {
            if (isConnectedToNetwork()!!) {
                webWToday?.reload()
                noConnectionLayoutVisibility(GONE)
            } else {
                noConnectionLayoutVisibility(VISIBLE)
            }
        }

    }

    private fun WtodayMainFragmentBinding.setClient() {

        webWToday?.webViewClient = object : WebViewClient() {
            @TargetApi(Build.VERSION_CODES.M)
            override fun onReceivedError(webView: WebView, request: WebResourceRequest, error: WebResourceError) {
                super.onReceivedError(webView, request, error)
                handleError(error.errorCode)
            }

            override fun onReceivedError(webView: WebView, errorCode: Int, description: String, url: String) {
                super.onReceivedError(webView, errorCode, description, url)
                handleError(errorCode)
            }
        }

        webWToday?.webChromeClient = object : WebChromeClient() {
            override fun onCreateWindow(view: WebView?, isDialog: Boolean, isUserGesture: Boolean, resultMsg: Message?): Boolean {
                val data = view?.hitTestResult?.extra

                try {
                    var intent: Intent?
                    when (data?.isEmailValid()) {
                        true -> Utils.sendEmail(data)
                        else -> data?.let {
                            val uri = Uri.parse(it)
                            intent = Intent(Intent.ACTION_VIEW, uri)
                            activity?.packageManager?.let { packageManager -> if (intent?.resolveActivity(packageManager) != null) startActivity(intent) }
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, e.message!!)
                }
                return false
            }
        }

    }

    fun scrollToTop() = binding.webWToday?.apply {
        ObjectAnimator.ofInt(this, "scrollY", scrollY, 0)?.run {
            setDuration(SCROLL_UP_ANIM_DURATION).start()
        }
    }

    private fun handleError(errorCode: Int) = when (errorCode) {
        ERROR_CONNECT, ERROR_TIMEOUT -> noConnectionLayoutVisibility(VISIBLE)
        else -> {
        }
    }


    private fun noConnectionLayoutVisibility(state: Int) {
        binding.incNoConnectionHandler?.root?.visibility = state
    }

    override fun onShowProductListing(categoryId: String, categoryName: String) {
        (activity as? BottomNavigationActivity)?.pushFragment(ProductListingFragment.newInstance(ProductsRequestParams.SearchType.NAVIGATE, categoryName, categoryId,
            isBrowsing = false,
            sendDeliveryDetails = false
        ))
    }

    override fun onAddIngredientsToShoppingList(ingredients: String) {
    }

    override fun onShowProductDetail(productId: String, skuId: String) {
        activity?.runOnUiThread { retrieveProduct(productId, skuId) }
    }

    override fun sendEventToFirebase(eventName: String, parameter: String) {
        val parameterName = eventName.replace("wtoday_", "")
        val arguments = HashMap<String, String>()
        arguments.put(parameterName, parameter)
        activity?.apply {Utils.triggerFireBaseEvents(eventName, arguments, this) }
    }

    override fun onResume() {
        super.onResume()
        activity?.let { Utils.setScreenName(it, FirebaseManagerAnalyticsProperties.ScreenNames.WTODAY) }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        when (hidden) {
            false -> (activity as? BottomNavigationActivity)?.hideToolbar()
            true -> cancelPDPRequest()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        cancelPDPRequest()
    }

    private fun cancelPDPRequest() = mGetProductDetail?.apply {
        if (isCanceled) {
            cancel()
        }
    }

    override fun progressBarVisibility(isDisplayed: Boolean) {
        binding.flProgressContainer?.visibility = if (isDisplayed) VISIBLE else GONE
    }


    /**
    Important: in future if any change required
    like currently WToday tab is default tab
    if default tab change then this notification Permission
     should be there.
     */
    private fun askNotificationPermission() {
        // This is only necessary for API level >= 33 (TIRAMISU)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(requireActivity(),
                    Manifest.permission.POST_NOTIFICATIONS) ==
                PackageManager.PERMISSION_GRANTED
            ) {
             // Do Nothing.... can post notifications.
            } else if (shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)) {
                notificationPermission.launchNotificationPermission()
            } else {
                // ask for the permission
                notificationPermission.launchNotificationPermission()
            }
        }
    }

    override fun permissionGranted(requestCode: Int) {

    }

}